central/complianceoperator/v2/compliancemanager/manager_impl.go

```diff
@@ -165,27 +169,96 @@ func (m *managerImpl) UpdateScanRequest(ctx context.Context, scanRequest *storag
 
+	return scanRequest, nil
+}
+
+// removeObsoleteResultsByClusters removes existing results related to removed clusters from scheduler configuration
+func (m *managerImpl) removeObsoleteResultsByClusters(ctx context.Context, oldScanConfig *storage.ComplianceOperatorScanConfigurationV2, newScanConfig *storage.ComplianceOperatorScanConfigurationV2) {
+	oldClusterIDs := set.NewStringSet()
+	for _, oldCluster := range oldScanConfig.GetClusters() {
+		oldClusterIDs.Add(oldCluster.GetClusterId())
+	}
+
+	newClusterIDs := set.NewStringSet()
+	for _, newCluster := range newScanConfig.GetClusters() {
+		newClusterIDs.Add(newCluster.GetClusterId())
+	}
+
+	removedClusterIDs := oldClusterIDs.Difference(newClusterIDs).AsSlice()
+	if len(removedClusterIDs) == 0 {
+		return
+	}
+
 	// Send delete to sensor for any clusters that were deleted
-	m.processClusterDelete(ctx, scanRequest, deletedClusters)
+	m.processClusterDelete(ctx, newScanConfig, removedClusterIDs)
 
-	return scanRequest, nil
+	err := m.resultsDS.DeleteResultsByScanConfigAndCluster(ctx, oldScanConfig.GetScanConfigName(), removedClusterIDs)
+	if err != nil {
+		log.Errorf("removing obsolete scan results for clusters %v: %v", removedClusterIDs, err)
+	}
+}
+
+// removeObsoleteResultsByProfiles removes existing results related to removed profiles from scheduler configuration
+func (m *managerImpl) removeObsoleteResultsByProfiles(ctx context.Context, oldScanConfig *storage.ComplianceOperatorScanConfigurationV2, newScanConfig *storage.ComplianceOperatorScanConfigurationV2) {
+	oldProfileNames := set.NewStringSet()
+	for _, oldProfile := range oldScanConfig.GetProfiles() {
+		oldProfileNames.Add(oldProfile.GetProfileName())
+	}
+
+	newProfileNames := set.NewStringSet()
+	for _, newProfile := range newScanConfig.GetProfiles() {
+		newProfileNames.Add(newProfile.GetProfileName())
+	}
+
+	removedProfileNames := oldProfileNames.Difference(newProfileNames).AsSlice()
+	if len(removedProfileNames) == 0 {
+		return
+	}
+
+	setRemovedProfileRuleNames := set.NewStringSet()
+	for _, profileName := range removedProfileNames {
+		query := search.NewQueryBuilder().AddExactMatches(search.ComplianceOperatorProfileName, profileName).ProtoQuery()
+		profiles, err := m.profileDS.SearchProfiles(ctx, query)
+		if err != nil {
+			log.Errorf("unable to get profile %q", profileName)
+			return
+		}
+
+		for _, profile := range profiles {
+			for _, rule := range profile.GetRules() {
+				setRemovedProfileRuleNames.Add(rule.GetRuleName())
+			}
+		}
+	}
+
+	oldClusters := oldScanConfig.GetClusters()
+	removedRuleRefIDs := make([]string, 0, len(oldClusters))
+	for _, oldCluster := range oldClusters {
+		for removedProfileRuleName := range setRemovedProfileRuleNames {
+			removedRuleRefIDs = append(removedRuleRefIDs, internaltov2storage.BuildNameRefID(oldCluster.GetClusterId(), removedProfileRuleName))
+		}
+	}
+
+	err := m.resultsDS.DeleteResultsByScanConfigAndRules(ctx, oldScanConfig.GetScanConfigName(), removedRuleRefIDs)
```

Comment:
I think this would be sightly more efficient if we use the `profilerefid` and `scanrefid`. With one scan you can delete multiple checkResults in one query. Also, profiles map to 1 or 2 scans but they can have a lot of rules. The flow could be: profile -> scan -> checkResult.